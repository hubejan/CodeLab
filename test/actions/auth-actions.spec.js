import { teacherLoginMock, teacherLogoutMock, error } from './async-utils/auth-mocks';

const mockedTeacherLogin = jest.fn(teacherLoginMock);
const mockedTeacherLogout = jest.fn(teacherLogoutMock);

describe('action creators', () => {
  // logout needs to be changed to reflect async auth revoke
  it('should create a LOGOUT action', () => {
    const dispatcherFunc = mockedTeacherLogout();
    const mockDispatch = jest.fn();
    const mockGetState = jest.fn();
    mockGetState.mockReturnValueOnce({ auth: { token: 'a valid token' } })
                .mockReturnValue({ auth: { token: 'invalid token' } });
                // different returns on first and second call to this func

    return dispatcherFunc(mockDispatch, mockGetState)
    .then(() => {
      expect(mockDispatch).toBeCalledWith({ type: 'LOGOUT' });
      expect(mockGetState).toHaveBeenCalled();
      return dispatcherFunc(mockDispatch, mockGetState);
    })
    .then(() => {
      expect(mockDispatch.mock.calls.length).toBe(1);
      expect(mockGetState.mock.calls.length).toBe(2);
      return expect(mockDispatch.mock.calls).toMatchSnapshot();
    });
  });

  it('should create an action with a valid token and username on successful login', () => {
    const dispatcherFunc = mockedTeacherLogin('a valid code');
    const mockDispatch = jest.fn();

    return dispatcherFunc(mockDispatch)
    .then(() => {
      expect(mockDispatch).toBeCalledWith({
        type: 'LOGIN_SUCCESS',
        token: 'a valid access token',
        username: 'Dummy Name'
      });
      return expect(mockDispatch.mock.calls).toMatchSnapshot();
    });
  });

  it('should create an action with an error object on failed login', () => {
    const dispatcherFunc = mockedTeacherLogin('INVALID CODE');
    const mockDispatch = jest.fn();

    return dispatcherFunc(mockDispatch)
    .then(() => {
      expect(mockDispatch).toBeCalledWith({ type: 'LOGIN_FAILURE', error });
      return expect(mockDispatch.mock.calls).toMatchSnapshot();
    });
  });
});
