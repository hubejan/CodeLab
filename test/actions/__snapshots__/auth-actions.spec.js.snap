// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action creators should create a LOGOUT action 1`] = `
Array [
  Array [
    Object {
      "type": "LOGOUT",
    },
  ],
]
`;

exports[`action creators should create an action with a valid token and username on successful login 1`] = `
Array [
  Array [
    Object {
      "token": "a valid access token",
      "type": "LOGIN_SUCCESS",
      "username": "Dummy Name",
    },
  ],
]
`;

exports[`action creators should create an action with an error object on failed login 1`] = `
Array [
  Array [
    Object {
      "error": Object {
        "error": "bad validation",
        "error_description": "incorrect code",
        "error_uri": "/error",
      },
      "type": "LOGIN_FAILURE",
    },
  ],
]
`;
